proposals/0024-repair-and-restart.md

```diff
@@ -0,0 +1,107 @@
+---
+simd: '0024'
+title: Automatially find confirmed slots and repair them before a cluster restart
+authors:
+  - Wen Xu (Solana Labs)
+category: Standard
+type: Core
+status: Draft
+created: 2023-04-07
+feature: (fill in with feature tracking issues once accepted)
+---
+
+## Summary
+
+Improve the current [cluster restart](https://docs.solana.com/running-validator/restart-cluster)
+procedure such that validators can automatically figure out confirmed slots,
+generate a snapshot, and then proceed to restart if everything looks fine.
+
+## Motivation
+Currently during a [cluster restart](https://docs.solana.com/running-validator/restart-cluster), validator operators need to decide latest optimistically confirmed slot, create a snapshot,
+then restart the validators with new commandline arguments.
+
+The current process involves a lot of human intenvention, if people make a
+mistake in deciding the latest optimistically confirmed slot, it could mean
+rollback of user transactions after they have been confirmed, which is not
+acceptable.
+
+We aim to automate the finding of oc slot and snapshot generation, so that
+we can lower the possibility of human mistakes in the cluster restart process.
+
+## Alternatives Considered
+
+See [Handling Common Solana Outages](https://docs.google.com/document/d/1RkNAyz-5aKvv5FF44b8SoKifChKB705y5SdcEoqMPIc/edit#bookmark=id.jtrjbe6g4mk3) for details.
+
+There are many proposals about automatically detecting that the cluster is
+in an outage so validators should enter a recovery process automatically.
+
+While getting human out of the loop greatly improves recovery speed,
+automaticlly restarting the whole cluster still seems risky. Because if
+the recovery process itself doesn't work, it might be some time before
+we can get human's attention. And it doesn't solve the cases where new binary
+is needed. So for now we still plan to have human in the loop.
+
+## New Terminology
+
+* SilentRepairMode - when validators restart in this new mode, they will
+talk with each other to find the highest oc slot and repair all necessary
+blocks. To improve speed and guarantee simplicity, Turbine, vote, and new
+block generation are paused in this mode.
+
+## Detailed Design
+
+A new command line arg --RepairAndRestart <optional_slots_to_send_on_last_voted_fork> is
```

Comment:
Actually you are right, chatted with Carl about this, he thinks we can just send an Epoch of slots, because currently the cluster is handling EpochSlots fine. I think it's probably enough to send 12 hours of slots by default, because normally you would have decided to restart within 12 hours during an outage. So right now I set it to 12 hours by default.